# Phase **2: Data Understanding**

In [2]:
import pandas as pd
df = pd.read_csv('Pakistan Largest Ecommerce Dataset.csv')
df.head()

C:\Users\Abdul Basit\AppData\Local\Temp\ipykernel_15400\3846755774.py:2: DtypeWarning: Columns (1,2,3,7,8,9,11,12,13,14,17,18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('Pakistan Largest Ecommerce Dataset.csv')


,item_id,status,created_at,sku,price,qty_ordered,grand_total,increment_id,category_name_1,sales_commission_code,...,Month,Customer Since,M-Y,FY,Customer ID,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24,Unnamed: 25
0,211131.0,complete,7/1/2016,kreations_YI 06-L,1950.0,1.0,1950.0,100147443,Women's Fashion,\N,...,7.0,2016-7,7-2016,FY17,1.0,NaN,NaN,NaN,NaN,NaN
1,211133.0,canceled,7/1/2016,kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...,240.0,1.0,240.0,100147444,Beauty & Grooming,\N,...,7.0,2016-7,7-2016,FY17,2.0,NaN,NaN,NaN,NaN,NaN
2,211134.0,canceled,7/1/2016,Ego_UP0017-999-MR0,2450.0,1.0,2450.0,100147445,Women's Fashion,\N,...,7.0,2016-7,7-2016,FY17,3.0,NaN,NaN,NaN,NaN,NaN
3,211135.0,complete,7/1/2016,kcc_krone deal,360.0,1.0,60.0,100147446,Beauty & Grooming,R-FSD-52352,...,7.0,2016-7,7-2016,FY17,4.0,NaN,NaN,NaN,NaN,NaN
4,211136.0,order_refunded,7/1/2016,BK7010400AG,555.0,2.0,1110.0,100147447,Soghaat,\N,...,7.0,2016-7,7-2016,FY17,5.0,NaN,NaN,NaN,NaN,NaN


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 26 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   item_id                584524 non-null  float64
 1   status                 584509 non-null  object 
 2   created_at             584524 non-null  object 
 3   sku                    584504 non-null  object 
 4   price                  584524 non-null  float64
 5   qty_ordered            584524 non-null  float64
 6   grand_total            584524 non-null  float64
 7   increment_id           584524 non-null  object 
 8   category_name_1        584360 non-null  object 
 9   sales_commission_code  447346 non-null  object 
 10  discount_amount        584524 non-null  float64
 11  payment_method         584524 non-null  object 
 12  Working Date           584524 non-null  object 
 13  BI Status              584524 non-null  object 
 14   MV                    584524 non-

In [4]:
df.describe()

,item_id,price,qty_ordered,grand_total,discount_amount,Year,Month,Customer ID,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24,Unnamed: 25
count,584524.000000,5.845240e+05,584524.000000,5.845240e+05,584524.000000,584524.000000,584524.000000,584513.000000,0.0,0.0,0.0,0.0,0.0
mean,565667.074218,6.348748e+03,1.296388,8.530619e+03,499.492775,2017.044115,7.167654,45790.511965,NaN,NaN,NaN,NaN,NaN
std,200121.173648,1.494927e+04,3.996061,6.132081e+04,1506.943046,0.707355,3.486305,34414.962389,NaN,NaN,NaN,NaN,NaN
min,211131.000000,0.000000e+00,1.000000,-1.594000e+03,-599.500000,2016.000000,1.000000,1.000000,NaN,NaN,NaN,NaN,NaN
25%,395000.750000,3.600000e+02,1.000000,9.450000e+02,0.000000,2017.000000,4.000000,13516.000000,NaN,NaN,NaN,NaN,NaN
50%,568424.500000,8.990000e+02,1.000000,1.960400e+03,0.000000,2017.000000,7.000000,42856.000000,NaN,NaN,NaN,NaN,NaN
75%,739106.250000,4.070000e+03,1.000000,6.999000e+03,160.500000,2018.000000,11.000000,73536.000000,NaN,NaN,NaN,NaN,NaN
max,905208.000000,1.012626e+06,1000.000000,1.788800e+07,90300.000000,2018.000000,12.000000,115326.000000,NaN,NaN,NaN,NaN,NaN


## **Project Strategy: Core Metrics & Business Questions**
**Goal:** To analyze Pakistan's largest e-commerce dataset to uncover consumer behavior, optimize payment workflows, and identify seasonal sales patterns.
## **1. Core Project Metrics**
To answer the specific questions, we will track these essential metrics:
| **Metric** | **Purpose** |
| :---: | :---: |
| **Category-wise Revenue** | To identify which product categories are the ""Best Sellers"" in terms of PKR value. |
| **Payment Success Ratio** | To compare 'Completed' vs 'Canceled' orders for each payment method. |
| **Cancellation Density** | To track which months (like November) or days have the highest order failure rates. |
| **Average Order Value** | To segment customers into 'Budget', 'Mid-range', and 'Premium' buyers. |
| **High-Ticket Completion Rate** | Specifically tracking the success rate of orders above a certain price threshold (e.g. >100,000 PKR). |
## **2. Key Analysis Questions**
We will answer these specific questions to derive actionable business recommendations:

* **Product Performance:** Which categories generate the most sales and volume?

* **Payment Reliability:** Which payment methods drive successful completions vs. high cancellations?

* **High-Value Order Audit:** Why do premium/expensive items have a near-zero completion rate?

* **Temporal Trends:** How does seasonality (Months/Days) impact order density?

* **Counter-Intuitive Patterns:** Identifying hidden behaviors, like the "Premium Transaction Paradox.

In [5]:
# Checking the shape of the dataset.
df.shape

(1048575, 26)

In [6]:
# Checking the columns of the dataset.
df.columns

Index(['item_id', 'status', 'created_at', 'sku', 'price', 'qty_ordered',
       'grand_total', 'increment_id', 'category_name_1',
       'sales_commission_code', 'discount_amount', 'payment_method',
       'Working Date', 'BI Status', ' MV ', 'Year', 'Month', 'Customer Since',
       'M-Y', 'FY', 'Customer ID', 'Unnamed: 21', 'Unnamed: 22', 'Unnamed: 23',
       'Unnamed: 24', 'Unnamed: 25'],
      dtype='object')

In [7]:
# Checking duplicate values in the dataset.
df.duplicated().sum()

np.int64(464050)

### **Data Integrity & Structure Check: Duplicates**
Before performing calculations, we must verify if the data contains redundant information or if "duplicates" actually represent valid business transactions.
#### **1. Row-Level Check**
* **Observation:** There are 464,050 duplicated rows in the dataset.

* **Analysis:** A significant portion of these duplicates are completely empty rows (Null values) that were likely added during the data extraction process.

* **Conclusion:** These rows do not represent unique transactions and must be removed in the Data Cleaning phase to ensure accurate results.
#### **2. Unnamed Columns & Noise**
* **Observation:** The dataset contains several unnecessary columns such as Unnamed: 21, Unnamed: 22, up to Unnamed: 25.

* **Impact:** These columns consume memory and provide no analytical value.

* **Action Plan:** These will be dropped during the cleaning process to streamline the dataframe.

In [8]:
# Calculating unique values for the Grain Table
total_rows = len(df)
unique_orders = df['increment_id'].nunique()
unique_customers = df['Customer ID'].nunique()
unique_products = df['sku'].nunique()

print(f"Total Rows: {total_rows}")
print(f"Unique Orders: {unique_orders}")
print(f"Unique Customers: {unique_customers}")
print(f"Unique Products (SKUs): {unique_products}")

Total Rows: 1048575
Unique Orders: 408785
Unique Customers: 115326
Unique Products (SKUs): 84889


## **Phase 1: Final Report (The Transaction Structure)**
### **Understanding the Transaction Structure (Data Grain)**
Before moving to cleaning, we analyzed the unique entities within our 1,048,575 rows to understand the "Grain" of our data:
| **Entity** | **Count** | **Business Implication** |
| :---: | :---: | :---: |
| **Total Rows** | 1,048,575 | Represents every single line item recorded in the system. |
| **Unique Orders** | 408,785 | On average, each order contains ~2.5 different products. |
| **Unique Customers** | 115,326 | Indicates a healthy customer base with potential repeat purchase behavior. |
| **Unique Products** | 84,889 | Represents a diverse catalog, requiring grouping by Category for insights. |

**Analysis Insight:** Since the number of Rows is significantly higher than Unique Orders, we must treat each row as a "Line Item" rather than a single "Order." This distinction is crucial for calculating the Average Order Value (AOV) later.


In [9]:
# Investigation of Unnamed Columns
unnamed_cols = [col for col in df.columns if 'Unnamed' in col]

for col in unnamed_cols:
    non_null_count = df[col].notnull().sum()
    unique_vals = df[col].dropna().unique()
    print(f"Column: {col}")
    print(f"- Non-Null Values: {non_null_count}")
    print(f"- Unique Values (if any): {unique_vals}")
    print("-" * 30)

Column: Unnamed: 21
- Non-Null Values: 0
- Unique Values (if any): []
------------------------------
Column: Unnamed: 22
- Non-Null Values: 0
- Unique Values (if any): []
------------------------------
Column: Unnamed: 23
- Non-Null Values: 0
- Unique Values (if any): []
------------------------------
Column: Unnamed: 24
- Non-Null Values: 0
- Unique Values (if any): []
------------------------------
Column: Unnamed: 25
- Non-Null Values: 0
- Unique Values (if any): []
------------------------------


### **Final Investigation: Unnamed Columns & Noise**
Before proceeding to data cleaning, a deep dive was conducted into the "Unnamed" columns (Unnamed: 21 to Unnamed: 25) to ensure no critical data is lost.
* **Observation:** The investigation revealed that all Unnamed columns have 0 non-null values and no unique data.

* **Result:** These columns are confirmed as "Data Noise" and will be removed in the next phase to optimize memory usage and clarity.

In [10]:
# Save the cleaned dataset for future use
df.to_csv('Cleaned_Pakistan_Ecommerce_Dataset.csv', index=False)